server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -251,8 +251,13 @@ protected void handleOperationRequest(final Request request, final TransportChan
     }
 
     protected void handlePrimaryRequest(final ConcreteShardRequest<Request> request, final TransportChannel channel, final Task task) {
-        new AsyncPrimaryAction(
-            request, new ChannelActionListener<>(channel, transportPrimaryAction, request), (ReplicationTask) task).run();
+        final Releasable releasable = checkPrimaryLimits(request.request);
```

Comment:
I'm confused. Isn't this only starting to do memory accounting after the request has started being processed on the write thread pool?

Reply:
Yeah. I should have clarified that I'm still working on (that aspect) of this PR. I'll add WIP.